### ISSUE TITLE
Runs for a bit then errors out

### ISSUE THREAD

**Issue by SenorSmartyPants** [terrelsa13/media_cleaner#13]
Running in debug mode I can see it is getting plenty of responses from my server, but then just dies. I can open this URL in a browser fine.

```
http://rockpi:8096/emby/Users/5c0c92019ab2454a8bdcc7aba9abcdd6/Items/c8fb495ac3ac99a7e886113e529c55b4?enableImages=False&api_key=APIKEY
Traceback (most recent call last):
  File "/usr/lib/python3.9/urllib/request.py", line 1346, in do_open
    h.request(req.get_method(), req.selector, req.data, headers,
  File "/usr/lib/python3.9/http/client.py", line 1279, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1325, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1274, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.9/http/client.py", line 1034, in _send_output
    self.send(msg)
  File "/usr/lib/python3.9/http/client.py", line 974, in send
    self.connect()
  File "/usr/lib/python3.9/http/client.py", line 945, in connect
    self.sock = self._create_connection(
  File "/usr/lib/python3.9/socket.py", line 823, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
  File "/usr/lib/python3.9/socket.py", line 954, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno -3] Try again

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/storage/media/media_cleaner.py", line 3391, in <module>
    deleteItems=get_items(cfg.server_url, cfg.user_keys, cfg.access_token)
  File "/storage/media/media_cleaner.py", line 2090, in get_items
    item_info=get_additional_item_info(server_url, user_key, item['Id'], auth_key, 'episode_item')
  File "/storage/media/media_cleaner.py", line 1064, in get_additional_item_info
    itemInfo=requestURL(url, cfg.DEBUG, lookupTopic, cfg.api_request_attempts)
  File "/storage/media/media_cleaner.py", line 1028, in requestURL
    with request.urlopen(url) as response:
  File "/usr/lib/python3.9/urllib/request.py", line 214, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.9/urllib/request.py", line 517, in open
    response = self._open(req, data)
  File "/usr/lib/python3.9/urllib/request.py", line 534, in _open
    result = self._call_chain(self.handle_open, protocol, protocol +
  File "/usr/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 1375, in http_open
    return self.do_open(http.client.HTTPConnection, req)
  File "/usr/lib/python3.9/urllib/request.py", line 1349, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [Errno -3] Try again>
```

**Comment by SenorSmartyPants** [comment]
It's in the gather info stage, no real output in none debug yet at this point.

**Comment by terrelsa13** [comment]
When you get a chance. Please Pastebin the contents of your configuration file so I can try to reproduce this on my end. You can x-out the access token.

**Comment by SenorSmartyPants** [comment]
https://pastebin.com/j2U1UkT5

And here's some stats on my JF server.

```
MovieCount: 315,
SeriesCount: 169,
EpisodeCount: 5119,
```

But I'm currently testing with only 1 tv library that I have 1 series in with 73 episodes.

**Comment by terrelsa13** [comment]
Will you include all information printed with `DEBUG=1`. There will be quite a bit of data returned from the API call that happened before this error. There should be episode data printed out before `http://rockpi:8096/emby/Users/5c0c92019ab2454a8bdcc7aba9abcdd6/Items/c8fb495ac3ac99a7e886113e529c55b4?enableImages=False&api_key=APIKEY`

I will want to see everything the script prints out to the command line. Again, you can x-out your access token. But there really is no need if Jellyfin is sitting safely behind your firewall.

**Comment by SenorSmartyPants** [comment]
Here's the full debug log with error.

https://file.io/WSSUsTV7tRhw

I ran it in debug=1 mode but with IP instead of DNS name. It finished this time. 

Why does it pull info on shows in a library that I don't even have it monitoring?

**Comment by terrelsa13** [comment]
Looks like the link to the debug file was deleted. At least that's what I see on my end.

I am curios about why the host name works for some of the info gathering API calls but not all of of them? Does the machine Jellyfin is running on have a host name and local domain (e.g. `.mynetwork`)? If you know your local domain try using it along with the host name (e.g. `http://rockpi.mynetwork`). Let me know if that allows the script to finish.

> Why does it pull info on shows in a library that I don't even have it monitoring?

Check out this option in the config file:
#### Control if the script will request metadata only for played media items OR played and not played media items:
```python
#----------------------------------------------------------#
#  0 - Request metadata only for played media items in monitored libraries
#   When single user, script will complete faster, no downside
#   When multiple users, script will complete faster BUT...
#   The script will only be able to keep a media item when a user has set it as a favorite and has played it
#  1 - Request metadata for played and not played media items in monitored libraries
#   When single user, script will complete slower, slower is the downside
#   When multiple users, script will complete slower BUT...
#   The script is able to keep a media item when a user has set it as a favortie but has not played it
# (1 : default)
#----------------------------------------------------------#
request_not_played=1
```
**Long story short;** for the script to work with multiple users it has to get the watched and favorite information for all media of a certain type. If you plan to only use this with a single user it will run much faster if you set `requested_not_played=0`.

**Comment by SenorSmartyPants** [comment]
I have multiple users. I have not tried with a local domain yet, but it would hit the same dns server (a pihole) so i'm not sure why that would make a difference.

But if a library isn't monitored for deletes at all, why not just include something like this
`ParentID=<IDoflibrary>&IsFavorite=true&recursive=true&includeItemTypes=series,season,episode&enableUserData=true`

That would get all favorite the tv related objects for the library that is monitored, instead of all tv stuff from all libraries.
If `IsFavorite` doesn't work on both jellyfin and emby, there's also `filters=IsFavorite`
https://api.jellyfin.org/#operation/GetItemsByUserId

**Comment by SenorSmartyPants** [comment]
Looking over my config again, I see that you are filtering on directory for block/allow lists, and not library IDs. So ParentID doesn't match that filtering parameter.

**Comment by terrelsa13** [comment]
> But if a library isn't monitored for deletes at all, why not just include something like this
ParentID=<IDoflibrary>&IsFavorite=true&recursive=true&includeItemTypes=series,season,episode&enableUserData=true

> That would get all favorite the tv related objects for the library that is monitored, instead of all tv stuff from all libraries.
If IsFavorite doesn't work on both jellyfin and emby, there's also filters=IsFavorite

This. Much appreciated! I had tried to get this to work a while back but couldn't. You are correct, this is the right way the script should be implemented. But I needed something in a hurry and just kept down the path of least resistance at the time. Thanks to your clue I was able to finally get this API call to work. I will use these to improve the script.

> Looking over my config again, I see that you are filtering on directory for block/allow lists, and not library IDs. So ParentID doesn't match that filtering parameter.

Correct. This script will get updated to use IDs instead of paths for libraries. All of the other filtering (artists, genres, network, etc...) is using IDs.

**Comment by SenorSmartyPants** [comment]
You're welcome. I've messed around with the JF API for item queries a lot. I'm open to questions if you have them

**Comment by SenorSmartyPants** [comment]
I do like the idea of being able to filter by directory,  but as you saw it's not easy to query that way.

**Comment by SenorSmartyPants** [comment]
http://192.168.1.4:8096/Users/5c0c92019ab2454a8bdcc7aba9abcdd6/Items/?locationTypes=FileSystem&Filters=IsPlayed&IncludeItemTypes=Episode&ParentID=71a73d06363d25418d8bb64613ea98f6&Recursive=true&SortBy=DatePlayed&SortOrder=Ascending&Limit=12&StartIndex=0&api_key=APIKEY&enableImages=false

Just thought I'd share this query. Gets all played local filesystem episodes of a parentID (could be library, series, or season) sorted by date played, oldest plays first, 12 episodes at a time. I also turned off the return of image data, since that does not seem relevant.

**Comment by SenorSmartyPants** [comment]
collapseBoxSetItems=false is useful on movies to make sure you are examining movies and not boxsets.

**Comment by terrelsa13** [comment]
I'll have to check if doing the top down query using the aggregate library Id also returns the same results as bottom up.

If it does I will have to do some restructuring. Before multi-user capabilities were added the structure got watched items and then checked up the tree to determine if it or any parents, genres, artists, etc... were favorited.

**Comment by terrelsa13** [comment]
Closing.
